velox/dwio/parquet/reader/TimestampColumnReader.h

```diff
@@ -65,14 +96,131 @@ class TimestampColumnReader : public IntegerColumnReader {
+          rows,
+          dwio::common::ColumnVisitor<
+              int128_t,
+              ParquetInt96TimestampRange,
+              ExtractValues,
+              isDense>(newRange, this, rows, extractValues));
+    } else {
+      reinterpret_cast<Reader*>(this)->Reader::readWithVisitor(
+          rows,
+          dwio::common::
+              ColumnVisitor<int128_t, TFilter, ExtractValues, isDense>(
+                  *reinterpret_cast<TFilter*>(filter),
+                  this,
+                  rows,
+                  extractValues));
+    }
+    return;
+  }
+
+  template <
+      typename Reader,
+      bool isDense,
+      bool kEncodingHasNulls,
+      typename ExtractValues>
+  void processTimestampFilter(
+      velox::common::Filter* filter,
+      ExtractValues extractValues,
+      const RowSet& rows) {
+    if (filter == nullptr) {
+      readHelperTimestamp<Reader, velox::common::AlwaysTrue, isDense>(
+          &dwio::common::alwaysTrue(), rows, extractValues);
+      return;
+    }
+
+    switch (filter->kind()) {
+      case velox::common::FilterKind::kAlwaysTrue:
+        readHelperTimestamp<Reader, velox::common::AlwaysTrue, isDense>(
+            filter, rows, extractValues);
+        break;
+      case velox::common::FilterKind::kIsNull:
+        if constexpr (kEncodingHasNulls) {
+          filterNulls<int64_t>(
+              rows,
+              true,
+              !std::
+                  is_same_v<decltype(extractValues), dwio::common::DropValues>);
+        } else {
+          readHelperTimestamp<Reader, velox::common::IsNull, isDense>(
+              filter, rows, extractValues);
+        }
+        break;
+      case velox::common::FilterKind::kIsNotNull:
+        if constexpr (
+            kEncodingHasNulls &&
+            std::is_same_v<decltype(extractValues), dwio::common::DropValues>) {
+          filterNulls<int64_t>(rows, false, false);
+        } else {
+          readHelperTimestamp<Reader, velox::common::IsNotNull, isDense>(
+              filter, rows, extractValues);
+        }
+        break;
+      case velox::common::FilterKind::kTimestampRange:
+        readHelperTimestamp<Reader, velox::common::TimestampRange, isDense>(
+            filter, rows, extractValues);
+        break;
+      default:
+        VELOX_UNREACHABLE();
+    }
+  }
+
+  template <typename Reader, bool kEncodingHasNulls>
+  void readTimestamp(const RowSet& rows) {
+    const bool isDense = rows.back() == rows.size() - 1;
+    velox::common::Filter* filter =
+        scanSpec_->filter() ? scanSpec_->filter() : &dwio::common::alwaysTrue();
+    if (scanSpec_->keepValues()) {
+      if (scanSpec_->valueHook()) {
+        if (isDense) {
+          processValueHook<Reader, true>(rows, scanSpec_->valueHook());
+        } else {
+          processValueHook<Reader, false>(rows, scanSpec_->valueHook());
+        }
+      } else {
+        if (isDense) {
+          processTimestampFilter<Reader, true, kEncodingHasNulls>(
```

Comment:
Can we not use `readCommon<TimestampColumnReader, true>(rows);` and specialize `readHelper<TimestampColumnReader... >` inside processFilter?

Reply:
> specialize readHelper<TimestampColumnReader... > inside processFilter

I suppose that would bring the parquet-specific 'TimestampColumnReader' in 'SelectiveIntegerColumnReader.h' under common module, which shall be avoided as prior comment pointed out. Please kindly correct me if I miss understand something.

Reply:
Thanks for making the test changes. I see Jimmy's comment above. He seemed to suggest using the ColumnVisitor instance directly. Do we need all the other paths in readTimestamp and processTimestampFilter? If we do, then this is not ideal as much code is being duplicated.

Reply:
@majetideepak I understand your point. Now we only keep a readHelper function and cast 'this' as 'Reader*' in processFilter to call the specific function. Would you take another look? Thank you.